Updates to Tracks onboarding events props - Take 2

vladolaru opened the pull request:
Follow up to #9015 

#### Changes proposed in this Pull Request

- We add the URL-based `source` prop to both Connect page view and Connect CTA click
- We ensure that the PHP-based redirects to Connect page keep the GET `from` param of the request
- We ensure that connect URLs keep the GET `from` param of the request
- Updates to source determination to account more for the `from` param value 

#### Testing instructions

* Checkout the PR branch on your local client install and build it by running `npm run build:client`
* Remove/reset any onboarded account you currently have so the Connect page is accessible
* Make sure you have the Tracks Vigilante (p7H4VZ-3cB-p2) Chrome extension installed and opened so you can monitor Tracks events
* Deactivate the WooPayments plugin
* Go to WooCommerce > Home and click on the "Get paid" task item.
* You should be redirected to the Connect page with `&from=WCADMIN_PAYMENT_TASK` in the URL
* In Tracks Vigilante, you should see a `wcadmin_page_view` event with the `source: wcadmin-payment-task` prop
* If you already have a working Jetpack connection:
  * install [the Jetpack Debug Helper plugin](https://github.com/Automattic/jetpack-debug-helper) (download the `trunk` zip files and install+activate on your local client installation)
  * activate the "Broken token Utilities" module and go to [its settings page](http://localhost:8082/wp-admin/admin.php?page=broken-token)
  * Store your current connection data by clicking on "Store these options"
  * Break your connection completely by clicking successively on: "Clear blog token", "Clear user token", "Clear the Primary user", "Clear the blog ID"
* Go to [the Connect page](http://localhost:8082/wp-admin/admin.php?page=wc-admin&path=%2Fpayments%2Fconnect&from=WCADMIN_PAYMENT_TASK), make sure you have `&from=WCADMIN_PAYMENT_TASK` in your URL, and that you see a "Connect your store" CTA button.
* Click on "Connect your store" and you should be redirected back to the Connect page but with an error about not being able to establish a connection on localhost
* The page URL should still include `&from=WCADMIN_PAYMENT_TASK`
* Fire up your Jurassic Tube by running `npm run tube:start` and access your local WP admin through it.
* Go to WooCommerce > Home, click on the "Get paid" (or "Get paid with WooCommerce") task item and your should see your Connect page with `&from=WCADMIN_PAYMENT_TASK` in the URL
* Connect your store to your WPCOM account and you should end up in the MOX flow with `&source=wcadmin-payment-task` in the URL
* Complete onboarding and make sure you can successfully onboard an account.
* Run `npm run tube:stop` to stop the tunneling
* Go to Jetpack Debug > Broken Token and click on "Restore from stored options" to get your Jetpack connection back as it was.

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

vladolaru committed:
Add source to Connect Tracks events

vladolaru committed:
Use the same source values

vladolaru committed:
Maintain from GET param in connect URL

vladolaru committed:
Maintain source when redirecting to connect page

vladolaru committed:
Include from when redirecting from onboarding flow to connect

vladolaru committed:
Carry over from GET param when redirecting to Connect page

vladolaru committed:
Blend in from when determining onboarding source

vladolaru committed:
Add changelog

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9036` or branch name `update/tracks-onboarding-events-props-take2` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=update/tracks-onboarding-events-props-take2) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 298523d5466edd87e74bb75530dca85bbacad25d
- Build time: 2024-07-02 09:20:59 UTC

_Note: the build is updated when a new commit is pushed to this PR._

vladolaru commented on includes/class-wc-payments-account.php:
```diff
@@ -992,6 +993,12 @@ public function maybe_handle_onboarding() {
 			update_option( 'wcpay_menu_badge_hidden', 'yes' );
 
 			if ( isset( $_GET['wcpay-connect-jetpack-success'] ) ) {
+				$test_mode        = isset( $_GET['test_mode'] ) && wc_clean( wp_unslash( $_GET['test_mode'] ) );
```
Moving this code higher prevents a possible undefined issue.

oaratovskyi commented:
Regarding testing instructions:
- In my case, both times on the homepage I clicked `Get paid with WooPayments`.
- I think we should also add `npm run tube:stop` after testing and restore Jetpack connection from stored options.

oaratovskyi commented on includes/class-wc-payments-redirect-service.php:
```diff
@@ -122,9 +122,10 @@ public function redirect_to_account_link( array $args ): void {
+	 *                     Will fall back to keeping the `from` parameter in the current request URL, if present.
 	 */
-	public function redirect_to_connect_page( ?string $error_message = null, string $from = '' ): void {
+	public function redirect_to_connect_page( ?string $error_message = null, string $from = null ): void {
```
```suggestion
	public function redirect_to_connect_page( ?string $error_message = null, ?string $from = null ): void {
```

oaratovskyi replied:
We should change the type to nullable if we passing `null` as default value

vladolaru replied:
Done in https://github.com/Automattic/woocommerce-payments/pull/9036/commits/b2f1af4b745ab68b2bbd9e42537c8531b2667bda

oaratovskyi commented on includes/class-wc-payments-redirect-service.php:
```diff
@@ -135,11 +136,16 @@ public function redirect_to_connect_page( ?string $error_message = null, string
 
-		if ( '' !== $from ) {
+		// If we were not given a source, try to get it from the request URL.
+		if ( ! isset( $from ) && isset( $_GET['from'] ) ) {
```
could you please add unit test for this?

vladolaru replied:
Done in https://github.com/Automattic/woocommerce-payments/pull/9036/commits/298523d5466edd87e74bb75530dca85bbacad25d

oaratovskyi reviewed (commented):
Regarding the changes, they work well 👏 
I tested everything according to the instructions 🎉 
The code looks almost good, I left a feedback in the comments, I'll be happy to approve once they're addressed!

vladolaru committed:
Use WP standard way of getting referer

vladolaru committed:
Type hint fix

vladolaru committed:
test: Replace deprecated PHPUnit method

vladolaru committed:
test: Use WP standard function to get admin URL for more resilient tests

vladolaru committed:
test: Add PHPUnit tests for redirect service redirect_to_connect_page

vladolaru commented:
Thank you for reviewing this @oaratovskyi !

> I think we should also add npm run tube:stop after testing and restore Jetpack connection from stored options.

I've updated the testing instructions.

> The code looks almost good, I left a feedback in the comments, I'll be happy to approve once they're addressed!

I've added PHPUnit tests and fixed the type hinting.

oaratovskyi reviewed (approved):
Thanks for addressing the feedback! :shipit:
